server/lib/adapters/mssql.js

```diff
@@ -0,0 +1,231 @@
+        if (reqError) {
+          return reject(reqError);
+        }
+      });
+
+      const rows = [];
+      let resultReturned = false;
+
+      request.on("row", (columns) => {
+        const row = {};
+        _.forEach(columns, (column) => {
+          row[column.metadata.colName] = column.value;
+        });
+        rows.push(row);
+      });
+
+      request.on("done", (rowCount) => { // eslint-disable-line consistent-return
+        if (_.isNumber(rowCount) && !resultReturned) {
+          resultReturned = true;
+          return resolve({
+            rows
+          });
+        }
+      });
+
+      request.on("doneInProc", (rowCount) => { // eslint-disable-line consistent-return
+        if (_.isNumber(rowCount) && !resultReturned) {
+          resultReturned = true;
+          return resolve({
+            rows
+          });
+        }
+      });
+
+      request.on("doneProc", (rowCount) => { // eslint-disable-line consistent-return
+        if (_.isNumber(rowCount) && !resultReturned) {
+          resultReturned = true;
+          return resolve({
+            rows
+          });
+        }
+      });
+
+      conn.execSql(request);
+    });
+  });
+}
+
+/**
+ * Creates a readable stream that contains the query result.
+ * @param {tedious.Connection} client The MS SQL client.
+ * @param {string} query The query to execute.
+ * @param {Object} options The options.
+ *
+ * @returns {Promise} A promise object that wraps a stream.
+ */
+export function streamQuery(client, query, options = {}) {
+  return new Promise((resolve, reject) => {
+    const confoptions = _.merge(client.config.options, options);
+    const conf = _.cloneDeep(client.config);
+    conf.options = confoptions;
+    const conn = new tedious.Connection(conf);
+
+    const streamOpts = {};
+    streamOpts.ObjectMode = true;
+    streamOpts.highWaterMark = 10;
+    const stream = new Readable(streamOpts);
+
+    let resultReturned = false;
+
+    stream._read = function () { // eslint-disable-line func-names
+      conn.on("connect", (err) => { // eslint-disable-line consistent-return
+        if (err) {
+          stream.emit("error", err);
+          return reject(err);
+        }
+
+        const request = new tedious.Request(query, (reqError) => { // eslint-disable-line consistent-return
+          if (reqError) {
+            stream.emit("error", reqError);
+            return reject(reqError);
+          }
+        });
+
+        request.on("row", (columns) => {
+          const row = {};
+          _.forEach(columns, (column) => {
+            row[column.metadata.colName] = column.value;
+          });
+          stream.push(JSON.stringify(row));
+        });
+
+        request.on("done", (rowCount) => { // eslint-disable-line consistent-return
+          if (_.isNumber(rowCount) && !resultReturned) {
+            resultReturned = true;
+            stream.push(null);
```

Comment:
hm, not 100% clear thing for me. Could we emit `end` or `close` event here instead of pushing null value to the stream?